src/typed-error.ts

```diff
@@ -0,0 +1,29 @@
+class BaseError {
+    constructor () {
+        Error.apply(this, arguments)
+    }
+}
+BaseError.prototype = Object.create(Error.prototype)
+
+export class TypedError extends BaseError {
+	public name: string
+	public message: string
+	public stack: string
+	constructor(message?: Error | string) {
+		super()
+		let err: Error
+		if (message instanceof Error) {
+			err = message
+		} else {
+			err = new Error(message)
+		}
+		err.name = (this.constructor as any).name
+		this.name = err.name
+		this.message = err.message
+		if (Error.captureStackTrace != null) {
+			Error.captureStackTrace(this, this.constructor)
+		} else if (err.stack != null) {
+			this.stack = err.stack
```

Comment:
Is it true that we only need the `err` object here to get the stack? The `name` can be received directly, as well as the message. If true we can only create it in the else clause, right?